workflow/metrics/controller_metrics.go

```diff
@@ -2,70 +2,94 @@ package metrics
 
 import (
 	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/tools/cache"
 
 	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
-	"github.com/argoproj/argo/workflow/util"
 )
 
 // legacyWorkflowCollector collects metrics about all workflows in the cluster
 type controllerCollector struct {
-	store util.WorkflowLister
+	informer cache.SharedIndexInformer
+	gauges   map[wfv1.NodePhase]prometheus.Gauge
```

Comment:
Do we still want controller-level Workflow counters (in addition to gauges)? I argue not to introduce them unless they are asked for.

Reply:
I've several I'd like us to add:

```
		insignificantPodChange:          prometheus.NewCounter(prometheus.CounterOpts{Namespace: argoNamespace, Subsystem: workflowsSubsystem, Name: "insignificant_pod_update", Help: "Number of insignificant pod updates"}),
		updatesPersisted:                prometheus.NewCounter(prometheus.CounterOpts{Namespace: argoNamespace, Subsystem: workflowsSubsystem, Name: "updates_persisted", Help: "Number of times we persisted a workflow update"}),
		updatesReapplied:                prometheus.NewCounter(prometheus.CounterOpts{Namespace: argoNamespace, Subsystem: workflowsSubsystem, Name: "updates_reapplied", Help: "Number of times we re-applied a workflow update. Ideally should always be zero."}),
		podResourceVersionRepeated:      prometheus.NewCounter(prometheus.CounterOpts{Namespace: argoNamespace, Subsystem: workflowsSubsystem, Name: "pod_resource_version_repeated", Help: "Number of pod updates had the same resource version as the old one"}),
		podsProcessed:                   prometheus.NewCounter(prometheus.CounterOpts{Namespace: argoNamespace, Subsystem: workflowsSubsystem, Name: "pods_processed", Help: "Number of pod updates processed"}),
		significantPodChange:            prometheus.NewCounter(prometheus.CounterOpts{Namespace: argoNamespace, Subsystem: workflowsSubsystem, Name: "significant_pod_update", Help: "Number of significant pod updates"}),
		workflowResourceVersionRepeated: prometheus.NewCounter(prometheus.CounterOpts{Namespace: argoNamespace, Subsystem: workflowsSubsystem, Name: "workflow_resource_version_repeated", Help: "Number of workflow updates that the same resource version as the old one"}),
		workflowsProcessed:              prometheus.NewCounter(prometheus.CounterOpts{Namespace: argoNamespace, Subsystem: workflowsSubsystem, Name: "workflows_processed", Help: "Number of workflow updates processed"}),
```

Reply:
This will go in a separate PR. See https://github.com/argoproj/argo/issues/2927